virttest/qemu_vm.py

```diff
@@ -1683,35 +1705,49 @@ def sort_key(dev):
             if params["machine_type"].startswith("pseries"):
                 hotpluggable_cpus *= vcpu_threads
             smp = smp or vcpu_maxcpus - hotpluggable_cpus
         else:
-            if smp == 0 or vcpu_sockets == 0:
-                vcpu_dies = vcpu_dies or 1
-                vcpu_cores = vcpu_cores or 1
-                vcpu_threads = vcpu_threads or 1
-                if smp == 0:
-                    vcpu_sockets = vcpu_sockets or 1
-                    smp = vcpu_cores * vcpu_threads * vcpu_dies * vcpu_sockets
-                else:
-                    vcpu_sockets = smp // (vcpu_cores * vcpu_threads * vcpu_dies) or 1
-            elif vcpu_dies == 0:
-                vcpu_cores = vcpu_cores or 1
-                vcpu_threads = vcpu_threads or 1
-                vcpu_dies = smp // (vcpu_sockets * vcpu_cores * vcpu_threads) or 1
-            elif vcpu_cores == 0:
-                vcpu_threads = vcpu_threads or 1
-                vcpu_cores = smp // (vcpu_sockets * vcpu_threads * vcpu_dies) or 1
+            if utils_qemu.get_qemu_version(qemu_binary)[0] not in \
+                    VersionInterval(SMP_PREFER_VERSION_SCOPE):
+                # Prefer sockets over cores before 6.2
+                if smp == 0 or vcpu_sockets == 0:
+                    vcpu_cores = vcpu_cores or 1
+                    vcpu_threads = vcpu_threads or 1
+                    if smp == 0:
+                        vcpu_sockets = vcpu_sockets or 1
+                        smp = vcpu_cores * vcpu_threads * vcpu_clusters * vcpu_dies * vcpu_sockets
+                    else:
+                        vcpu_sockets = smp // (vcpu_cores * vcpu_threads * vcpu_clusters * vcpu_dies) or 1
+                elif vcpu_cores == 0:
+                    vcpu_threads = vcpu_threads or 1
+                    vcpu_cores = smp // (vcpu_sockets * vcpu_threads * vcpu_clusters * vcpu_dies) or 1
             else:
-                vcpu_threads = smp // (vcpu_cores * vcpu_sockets * vcpu_dies) or 1
+                # Prefer cores over sockets since 6.2
+                if smp == 0 or vcpu_cores == 0:
+                    vcpu_sockets = vcpu_sockets or 1
+                    vcpu_threads = vcpu_threads or 1
+                    if smp == 0:
+                        vcpu_cores = vcpu_cores or 1
+                        smp = vcpu_cores * vcpu_threads * vcpu_clusters * vcpu_dies * vcpu_sockets
+                    else:
+                        vcpu_cores = smp // (vcpu_sockets * vcpu_threads * vcpu_clusters * vcpu_dies) or 1
+                elif vcpu_sockets == 0:
+                    vcpu_threads = vcpu_threads or 1
+                    vcpu_sockets = smp // (vcpu_cores * vcpu_threads * vcpu_clusters * vcpu_dies) or 1
+
+            if vcpu_threads == 0:
+                # Try to calculate omitted threads at last
+                vcpu_threads = smp // (vcpu_cores * vcpu_sockets * vcpu_clusters * vcpu_dies) or 1
```

Comment:
Here you just follow a part of qemu design. Which your design, this prefers smp first, you're still using the old qemu design.

To use the new design, in this `else`, when `smp = 0`, all values should be: https://gitlab.com/qemu-project/qemu/-/blob/master/hw/core/machine-smp.c#L115-118

when `smp != 0`, "prefer socket" and "prefer core" should be: https://gitlab.com/qemu-project/qemu/-/blob/master/hw/core/machine-smp.c#L120-151

The old qemu design can refer to: https://github.com/qemu/qemu/blob/f3b8f18ebf344ab359e8f79f6ed777e740dae77c/hw/i386/pc.c#L1561-L1586=

Reply:
I see what you mean, actually my designs are equivalent.
finally:
https://gitlab.com/qemu-project/qemu/-/blob/master/hw/core/machine-smp.c#L150

https://github.com/avocado-framework/avocado-vt/pull/3416/files#L1728
and 
https://github.com/avocado-framework/avocado-vt/pull/3416/files#L1741

Reply:
After discussion, we decided to adjust the structure to be exactly the same as the source code. 
update later